text/0079-cdk-2.0.md

```diff
@@ -0,0 +1,337 @@
+```
+
+will automatically be changed to:
+
+```typescript
+import { App, Stack } from 'aws-cdk-lib';
+import * as s3 from 'aws-cdk-lib/s3';
+import { Cluster, Service } from 'aws-cdk-lib/ecs';
+```
+
+It is not yet known if imports will require change when using other languages. See [namespace renaming]() for more details regarding that.
+
+#### [Migration Guide]()
+
+A comprehensive guide for migrating cdk apps from v1 to v2 will be added to the aws-cdk documentation. A place for future migration guides will be defined and a markdown document will be written detailing what steps users should take when upgrading to cdk v2.
+
+This guide will detail:
+1. Installing `aws-cdk-lib` and removing `@aws-cdk/` modules.
+2. Upgrading version of `aws-cdk` and other toolchain packages to version `2.x`.
+3. Rewriting import statements in source code if necessary, and how to use the [import rewriter]()
+4. Removing usage of deprecated APIs.
+  - List of APIs being deprecated and links to the suggested alternatives for each.
+5. Removing references to old feature flags
+  - List of feature flags being removed and strategy for migrating to new behavior for each.
+
+### Additional Candidates for Inclusion:
+
+The following are changes whose implementation's have been identified as requiring a breaking change. Inclusion in the v2 release is up for debate. Arguments to include a specific feature should take into account any increase in cost to the development team required to ship the release, and to users required to migrate their apps to the new version. Though we want to keep both of these costs as low as possible, a breaking release is something that happens relatively infrequently, so the opportunity should be taken to include features determined to have a high payoff to users.
+
+In addition, since the v1 modules will continue to be supported for some time, and the `aws-cdk-lib` package will be made of up these modules, the inclusion of any of these candidate features would need to be implemented in a way that didn't affect the v1 modules. How this would be done has not been determined, but may be affected by the nature of any particular change.
+
+#### Recommended
+
+The following are good candidates for inclusion in this major version release
+
+- [Easier Identification of Experimental Modules and APIs](https://github.com/aws/aws-cdk-rfcs/issues/116)
+
+There have been long running discussions about separating experimental APIs into a separate module to make them a more explicit "opt in" feature. Users would either have to install a different module (i.e. `aws-cdk-lib-exerimental`) or import from a different entrypoint (`import * as s3 from 'aws-cdk-lib/experimental`). Since this could be performed at build time and be an optional setting thats turned on or off, it could be introduced in the 2.0 release without affecting the v1 modules, this seems like an ideal time to implement. The technical details of how to achieve this are yet to be determined.
+
```

Comment:
what about stable APIs that extend/use experimental APIs?

Reply:
Yeah we have to decide if that counts as stable or experimental for these purposes. Experimental feels right but does that remove too much useful stuff for the user? We should run some experiments to get metrics for how frequent stable apis calls experimental ones.

Reply:
By definition those APIs are stable (and jsii-diff will shout if you try to change them). I think we should add a validation to the jsii compiler that prohibits this (and switch all of those APIs to stable so we can enable this check from now on)

Reply:
* Using: it's *super* useful to be able to add an experimental field to a stable struct. If we say that stable functions are not allowed to take structs with experimental members in them, we close this useful avenue. I don't think we should take the route of prohibiting that.

There is something to be said for the struct *itself* not being experimental, but given that the struct itself can be stable and literally every field in it could still be experimental, I'm not sure how useful it would be to have that additional restriction.

* Extending: I was all ready to say that an API class/interface not be able to extend an experimental class/interface, but after thinking about the previous issue I'm not so sure about that anymore. Inheriting from an experimental base just means that you'll inherit some experimental members. It should behave somewhat the same as if you just hand-declared a bunch of experimental members, and that shouldn't be disallowed, so why should this?

I guess there's something to be said about the stability of the class/interface name.

But on the other hand, if an interface of a class `ISomething` is experimental, then a consumer shouldn't be declaring variables of type `ISomething` to store that class instance on. Declare variables to be `const x: SomeConcreteSublass = gimmeObject();` instead.

All in all, I'm not so sure I have a problem with this anymore.

To my mind, feels like in the "stable" rendition of the library, jsii should behave as if the experimental types weren't `export`ed (same as it does now for private base classes and private base interfaces).

Reply:
What I'm more concerned about is the naming of types.

Types should have the same FQNs whether they are experimental or stable. If we don't have that, switching a class from experimental to stable breaks all consumers, and that would be shit.